Info + Bulk Format APIs

davedelong opened the pull request:
This adds two new endpoints to the website:

- `GET /info.json` - returns a simple structure listing supported locales, calendars, time zones, and the tzdb version
- `POST /format.json` - given an array of requests to format timestamps in certain locales, calendars, and/or time zones with varying formats, this returns the formatted strings. The full details of the format are in the commit message.

(This also updates the packages to their latest versions)

davedelong committed:
Add new bulk formatting endpoint

`/format.json` takes a JSON array of format requests. These requests MUST have a `format` key, whose value is a Format object.

A Format object must have either:
- a `date` key with a value of `full`, `long`, `medium`, or `short`,
- a `time` key with one of the four values
- both a `date` and a `time` key with one of the four values
- a `template` key with a string value
- a `raw` key with a string value
Any other keys will cause a `400 Bad Request`

The format request may also have:
- a `locale` key with a valid locale identifier. If missing, it assumes `en_US_POSIX`
- a `calendar` key with a valid calendar identifier. If missing, it uses the locale's calendar, and falls back to `gregorian`
- a `timeZone` key with a valid time zone identifier. If missing, it uses the locale's time zone, and falls back to `GMT`
- a `timestamp` key whose value is either a double (seconds from 1970). If missing, it assumes "the current time"
- an `id` key, which is a client-supplied value
If the provided `locale`, `calendar`, or `timeZone` values do not match known values, a `400 Bad Request` will be returned.

Each response in the returned array always has:
- an `id`, which is whatever `id` value was provided in the request
- a `calendar` string, which is the identifier of the used calendar
- a `timeZone` string, which is the identifier of the used time zone
- a `locale` string, which is the identifier of the used locale
- a `format` string, which is the fully expanded date format string used
- a `timestamp` double, which is the timestamp (seconds from 1970) used for formatting
- a `value` string, which is the formatted date string

davedelong committed:
Add an `info.json` endpoint

Provides information about supported locales, calendars, and time zones

davedelong commented:
Well, what do you know! This would address #10 and #11

subdigital commented on Sources/App/Controllers/FormatAPIController.swift:
```diff
@@ -0,0 +1,222 @@
+        let calendarID: String
+        let timeZone: TimeZone
+        
+        if let localeID = request.locale {
+            let loweredLocaledID = localeID.lowercased()
+            guard let actualLocaleID = Locale.availableIdentifiers.first(where: { $0.lowercased() == loweredLocaledID }) else {
+                throw Abort(.badRequest, reason: "Unknown locale: '\(localeID)'")
+            }
+            locale = Locale(identifier: actualLocaleID)
+        } else {
+            locale = Locale(identifier: "en_US_POSIX")
+        }
```
perhaps cleaner to extract this and the time zone parsing / defaults into a separate method?

```swift
private func resolveLocale(_ request: FormatRequest) throws -> Locale {
    if let localeID = request.locale {
            let loweredLocaledID = localeID.lowercased()
            guard let actualLocaleID = Locale.availableIdentifiers.first(where: { $0.lowercased() == loweredLocaledID }) else {
                throw Abort(.badRequest, reason: "Unknown locale: '\(localeID)'")
            }
            locale = Locale(identifier: actualLocaleID)
        } else {
            locale = Locale(identifier: "en_US_POSIX")
        }
}
```

Then the callsite becomes pretty clean:

```swift
let locale = try resolveLocale(request)
let timeZone = try resolveTimeZone(request) // e.g.
```

subdigital commented on Sources/App/Controllers/FormatAPIController.swift:
```diff
@@ -0,0 +1,222 @@
+        } else {
+            calendar = locale.calendar
+            calendarID = "\(calendar.identifier)"
+        }
```
same for calendar

subdigital commented on Sources/App/Controllers/FormatAPIController.swift:
```diff
@@ -0,0 +1,222 @@
+        
+        let formatted = formatter.string(from: date)
+        
+        return FormatResponse(id: request.id,
+                              calendar: calendarID,
+                              timeZone: timeZone.identifier,
+                              locale: locale.identifier,
+                              timestamp: date.timeIntervalSince1970,
+                              format: formatter.dateFormat,
+                              value: formatted)
+    }
```
```suggestion
        return FormatResponse(
            id: request.id,
            calendar: calendarID,
            timeZone: timeZone.identifier,
            locale: locale.identifier,
            timestamp: date.timeIntervalSince1970,
            format: formatter.dateFormat,
            value: formatted
        )
    }
```

minor formatting nit

subdigital commented on Sources/App/Controllers/FormatAPIController.swift:
```diff
@@ -0,0 +1,222 @@
+    }
+}
+
+struct FormatRequest: Content {
```
Could you add some docs to this type to include some example usages?

davedelong commented on Sources/App/Controllers/FormatAPIController.swift:
```diff
@@ -0,0 +1,222 @@
+            
+            if allKeys == [.date] {
+                self = .date(try container.decode(Style.self, forKey: .date))
+            } else if allKeys == [.date, .time] || allKeys == [.time, .date] {
```
That was how I had originally written it, but I took it out because I figured the combinations were so minimal (only 6 possible variations, and never an array longer than 2 values) that it wasn't worth the extra CPU time to turn things into sets for "proper" matching.

I'm happy to put it back in if you're prefer it that way.

davedelong committed:
Address feedback

davedelong committed:
Restore proper capitalization of calendar identifiers

When I made the calendar lookup case insensitive, I overlooked that this would affect the output of the `/info.json` response.

This restores the "proper" capitalization of those identifiers and then does the case-insensitive lookup in a manner similar to time zones and locales

davedelong committed:
Update hard-coded calendar identifiers

Make sure the calendar identifiers match the CLDR identifiers used by ICU

subdigital reviewed (changes requested):
I was able to give this a run through and it looks good. The last bit is the format API currently requires an array but the doc/examples show a single format request.

I think either we should update the docs to include this, or be slightly more lenient in what we accept (first try to decode `[FormatJSONRequest].self` and fall back to `FormatJSONRequest.self`).

davedelong committed:
Address feedback around documentation

davedelong commented:
> I think either we should update the docs to include this, or be slightly more lenient in what we accept (first try to decode `[FormatJSONRequest].self` and fall back to `FormatJSONRequest.self`).

I updated the documentation. It's not immediately obvious to me how I could make the route method return either a `[FormatResponse]` or a `FormatResponse` depending on what it manages to decode. Maybe it's a second method but the same route…?

davedelong commented:
> Maybe it's a second method but the same route…?

yep, that's exactly what it is! @subdigital I can add that too if you'd prefer.

davedelong commented:
This is in now, so it can be closed. :)
